embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/capture/crumbs/FragmentBreadcrumbDataSource.kt

```diff
@@ -1,90 +1,87 @@
 package io.embrace.android.embracesdk.capture.crumbs
 
-import io.embrace.android.embracesdk.arch.DataCaptureService
+import io.embrace.android.embracesdk.arch.datasource.NoInputValidation
+import io.embrace.android.embracesdk.arch.datasource.SpanDataSourceImpl
+import io.embrace.android.embracesdk.arch.datasource.startSpanCapture
+import io.embrace.android.embracesdk.arch.destination.StartSpanData
+import io.embrace.android.embracesdk.arch.destination.StartSpanMapper
+import io.embrace.android.embracesdk.arch.limits.UpToLimitStrategy
 import io.embrace.android.embracesdk.config.ConfigService
 import io.embrace.android.embracesdk.internal.clock.Clock
+import io.embrace.android.embracesdk.internal.spans.SpanService
 import io.embrace.android.embracesdk.logging.InternalEmbraceLogger
 import io.embrace.android.embracesdk.logging.InternalStaticEmbraceLogger
 import io.embrace.android.embracesdk.payload.FragmentBreadcrumb
-import io.embrace.android.embracesdk.utils.filter
-import java.util.Collections
+import io.embrace.android.embracesdk.spans.EmbraceSpan
 
 /**
  * Captures fragment breadcrumbs.
  */
 internal class FragmentBreadcrumbDataSource(
     private val configService: ConfigService,
     private val clock: Clock,
-    private val store: BreadcrumbDataStore<FragmentBreadcrumb> = BreadcrumbDataStore {
-        configService.breadcrumbBehavior.getFragmentBreadcrumbLimit()
-    },
+    spanService: SpanService,
     private val logger: InternalEmbraceLogger = InternalStaticEmbraceLogger.logger
-) : DataCaptureService<List<FragmentBreadcrumb>> by store {
+) : SpanDataSourceImpl(
+    spanService,
+    UpToLimitStrategy({ configService.breadcrumbBehavior.getFragmentBreadcrumbLimit() })
+),
+    StartSpanMapper<FragmentBreadcrumb> {
 
-    companion object {
+    private val fragmentSpans: MutableMap<String, EmbraceSpan> = mutableMapOf()
 
-        /**
-         * The default limit for how many open tracked fragments are allowed, which can be overridden
-         * by [RemoteConfig].
-         */
-        private const val DEFAULT_VIEW_STACK_SIZE = 20
-    }
-
-    internal val fragmentStack: MutableList<FragmentBreadcrumb> = Collections.synchronizedList(ArrayList<FragmentBreadcrumb>())
-
-    fun startFragment(name: String?): Boolean {
-        if (name == null) {
-            return false
-        }
-        synchronized(this) {
-            if (fragmentStack.size >= DEFAULT_VIEW_STACK_SIZE) {
-                return false
+    /**
+     * Called when a fragment is started.
+     */
+    fun startFragment(name: String?): Boolean = captureSpanData(
```

Comment:
Does this assume that the names of the fragment can never repeat? If we never have to show it, does it make sense to get the hashcode of the instance instead to identify it in the map - or something that is unique to that instance of the fragment?

Reply:
The developer ultimately supplies any names that come down to this level via our API, so there's not much we can do (short of maybe logging warnings)